File: packages/ios/Sources/Components/DirectionStep/DirectionStep.swift
import SwiftUI

/// What a step of a route asks for. The four turns, and the transitions the
/// routing engines describe: a level change by lift, escalator or stairs —
/// up or down — or by something unnamed; a same-level transition, a
/// walkway or a corridor to another building; and turning back. Each
/// platform draws the closest glyph its own icon set has, and the
/// instruction's words carry the rest.
public enum DirectionType: Sendable, CaseIterable {
    case straight, left, right, destination
    case liftUp, liftDown
    case escalatorUp, escalatorDown
    case stairsUp, stairsDown
    /// A level change by a transition the route does not name.
    case levelUp, levelDown
    /// A transition on the same level: a walkway, a corridor, another building.
    case transition
    case turnBack

    /// SF Symbols has no lift and no escalator: those, and an unnamed level
    /// change, show the direction of travel; stairs have their own figure.
    var iconName: String {
        switch self {
        case .straight: return "arrow.up"
        case .left: return "arrow.turn.up.left"
        case .right: return "arrow.turn.up.right"
        case .destination: return "mappin.and.ellipse"
        case .liftUp, .escalatorUp, .levelUp: return "arrow.up.to.line"
        case .liftDown, .escalatorDown, .levelDown: return "arrow.down.to.line"
        case .stairsUp, .stairsDown: return "figure.stairs"
        case .transition: return "arrow.forward.to.line"
        case .turnBack: return "arrow.uturn.backward"
        }
    }
}

public struct KozmosDirectionStep: View {
    let type: DirectionType
    let instruction: String
    let distance: String?
    let duration: String?
    
    public init(type: DirectionType, instruction: String, distance: String? = nil, duration: String? = nil) {
        self.type = type
        self.instruction = instruction
        self.distance = distance
        self.duration = duration
    }

    /// What VoiceOver reads for a step: the instruction, then the distance
    /// and the duration, as one element. The arrow says nothing the
    /// instruction does not, and a system symbol left audible reads its own
    /// description — "Up", or "Remove Map Pin" for the destination.
    static func accessibilityDescription(instruction: String, distance: String?, duration: String?) -> String {
        [instruction, distance, duration].compactMap { $0 }.filter { !$0.isEmpty }.joined(separator: ", ")
    }
    
    public var body: some View {
        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing150) {
            Circle()
                .fill(KozmosColors.primitivesColorsTheme500.opacity(0.1))
                .frame(width: KozmosDimensions.primitivesLayoutSizing500, height: KozmosDimensions.primitivesLayoutSizing500)
                .overlay(
                    Image(systemName: type.iconName)
                        .foregroundColor(KozmosColors.primitivesColorsTheme500)
                )
                .accessibilityHidden(true)
            
            VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing25) {
                Text(instruction)
                    .font(KozmosTypography.body)
                    .fontWeight(.medium)
                
                if let dist = distance {
                    Text(dist + (duration != nil ? " • \(duration!)" : ""))
                        .font(KozmosTypography.caption)
                        .foregroundColor(KozmosColors.primitivesColorsForeground500)
                }
            }
            Spacer()
        }
        .padding()
        .background(KozmosColors.primitivesColorsBackground0)
        .cornerRadius(KozmosDimensions.primitivesLayoutSpacing150)
        .overlay(
            RoundedRectangle(cornerRadius: KozmosDimensions.primitivesLayoutSpacing150)
                .stroke(KozmosColors.primitivesColorsBackground300, lineWidth: 1)
        )
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Self.accessibilityDescription(instruction: instruction, distance: distance, duration: duration))
    }
}

File: packages/android/src/main/java/com/kozmos/components/DirectionStep/DirectionStep.kt
package com.kozmos.components.directionstep

import com.kozmos.tokens.KozmosDimensions

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowUpward
import androidx.compose.material.icons.filled.ArrowDownward
import androidx.compose.material.icons.filled.Elevator
import androidx.compose.material.icons.filled.Escalator
import androidx.compose.material.icons.filled.Stairs
import androidx.compose.material.icons.filled.UTurnLeft
import androidx.compose.material.icons.filled.ArrowRightAlt
import androidx.compose.material.icons.filled.LocationOn
import androidx.compose.material.icons.filled.TurnLeft
import androidx.compose.material.icons.filled.TurnRight
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosThemeTokens

/**
 * What a step of a route asks for. The four turns, and the transitions the
 * routing engines describe: a level change by lift, escalator or stairs — up
 * or down — or by something unnamed; a same-level transition, a walkway or a
 * corridor to another building; and turning back. Each platform draws the
 * closest glyph its own icon set has, and the instruction's words carry the
 * rest.
 */
enum class DirectionType {
    Straight, Left, Right, Destination,
    LiftUp, LiftDown,
    EscalatorUp, EscalatorDown,
    StairsUp, StairsDown,
    /** A level change by a transition the route does not name. */
    LevelUp, LevelDown,
    /** A transition on the same level: a walkway, a corridor, another building. */
    Transition,
    TurnBack
}

/**
 * The arrow for a direction, one table for every part that draws one.
 * Turn icons must NOT auto-mirror: "turn left" stays a physical left turn
 * in RTL locales. Only reading-order affordances (back, forward, chevrons)
 * belong to Icons.AutoMirrored. Material has a lift, an escalator and
 * stairs of its own, with no direction: the instruction says up or down; an
 * unnamed level change shows the direction of travel.
 */
fun DirectionType.icon(): ImageVector = when (this) {
    DirectionType.Straight -> Icons.Default.ArrowUpward
    DirectionType.Left -> Icons.Default.TurnLeft
    DirectionType.Right -> Icons.Default.TurnRight
    DirectionType.Destination -> Icons.Default.LocationOn
    DirectionType.LiftUp, DirectionType.LiftDown -> Icons.Default.Elevator
    DirectionType.EscalatorUp, DirectionType.EscalatorDown -> Icons.Default.Escalator
    DirectionType.StairsUp, DirectionType.StairsDown -> Icons.Default.Stairs
    DirectionType.LevelUp -> Icons.Default.ArrowUpward
    DirectionType.LevelDown -> Icons.Default.ArrowDownward
    DirectionType.Transition -> Icons.Default.ArrowRightAlt
    DirectionType.TurnBack -> Icons.Default.UTurnLeft
}

@Composable
fun KozmosDirectionStep(
    type: DirectionType,
    instruction: String,
    modifier: Modifier = Modifier,
    distance: String? = null,
    duration: String? = null
) {
    val icon = type.icon()

    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(KozmosDimensions.semanticsRadiusControl))
            .background(KozmosThemeTokens.primitivesColorsBackground0)
            .border(1.dp, KozmosThemeTokens.primitivesColorsBackground300, RoundedCornerShape(KozmosDimensions.semanticsRadiusControl))
            .padding(KozmosDimensions.primitivesLayoutSpacing150),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Box(
            modifier = Modifier
                .size(KozmosDimensions.primitivesLayoutSizing500)
                .clip(CircleShape)
                .background(KozmosThemeTokens.primitivesColorsTheme500.copy(alpha = 0.1f)),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = KozmosThemeTokens.primitivesColorsTheme500,
                modifier = Modifier.size(KozmosDimensions.primitivesLayoutSizing300)
            )
        }
        
        Spacer(modifier = Modifier.width(KozmosDimensions.primitivesLayoutSpacing150))
        
        Column {
            Text(
                text = instruction,
                style = MaterialTheme.typography.titleMedium,
                color = KozmosThemeTokens.primitivesColorsForeground100
            )
            if (distance != null || duration != null) {
                Text(
                    text = "${distance ?: ""} ${if (duration != null) "• $duration" else ""}",
                    style = MaterialTheme.typography.bodySmall,
                    color = KozmosThemeTokens.primitivesColorsForeground500
                )
            }
        }
    }
}
